primitives/src/merkle_tree/field_based_mht/path.rs

```diff
@@ -0,0 +1,380 @@
+
+        leaf_index
+    }
+}
+
+impl<T: FieldBasedMerkleTreeParameters> PartialEq for FieldBasedMHTPath<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.path == other.path
+    }
+}
+
+impl<T: FieldBasedMerkleTreeParameters> FieldBasedMerkleTreePath for FieldBasedMHTPath<T> {
+    type H = T::H;
+
+    /// A Merkle Path for a leaf of a Merkle Tree with arity >= 2 will be made up of couples of nodes
+    /// and an integer. The nodes are all the siblings of the leaf ( in number MERKLE_ARITY - 1 )
+    /// and the integer is the position of the leaf, among its siblings, in the input of the hash
+    /// function (valid values are from 0 to MERKLE_ARITY - 1).
+    type Path = Vec<(Vec<<Self::H as FieldBasedHash>::Data>, usize)>;
+
+    type Parameters = T;
+
+    fn new(path: Self::Path) -> Self {
+        Self { path }
+    }
+
+    fn verify_without_length_check(
+        &self,
+        leaf: &<Self::H as FieldBasedHash>::Data,
+        root: &<Self::H as FieldBasedHash>::Data
+    ) -> Result<bool, Error> {
+
+        // Rate may also be smaller than the arity actually, but this assertion
+        // is reasonable and simplify the design. Should be also enforced by the
+        // MerkleTree that creates this instance, but let's do it again.
+        assert_eq!(<<Self::H as FieldBasedHash>::Parameters as FieldBasedHashParameters>::R, T::MERKLE_ARITY);
+
+        let mut digest = <Self::H as FieldBasedHash>::init_constant_length(T::MERKLE_ARITY, None);
+        let mut prev_node = leaf.clone();
+        for (sibling_nodes, position) in self.path.as_slice() {
+            assert_eq!(sibling_nodes.len(), T::MERKLE_ARITY - 1);
+
+            // Get the position of the node among its siblings
+            let prev_node_position = *position % T::MERKLE_ARITY;
+
+            // Update the digest respecting the position of each sibling
+            for i in 0..T::MERKLE_ARITY {
+                if i == prev_node_position {
+                    digest.update(prev_node.clone());
+                } else {
+                    let index = i % (T::MERKLE_ARITY - 1); // Be sure to not overflow the siblings vector
+                    digest.update(sibling_nodes[index]);
+                }
+            }
+
+            // Compute the parent node
+            prev_node = digest.finalize().unwrap();
+            digest.reset(None);
+        }
+
+        // Check final computed node is equal to the root
+        if prev_node == root.clone() {
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    fn get_raw_path(&self) -> Self::Path {
+        self.path.clone()
+    }
+
+    fn get_length(&self) -> usize {
+        self.path.len()
+    }
+}
+
+impl<T: FieldBasedMerkleTreeParameters> ToBytes for FieldBasedMHTPath<T> {
+    fn write<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        (self.path.len() as u8).write(&mut writer)?;
```

Comment:
instead of storing the sizes, we can derive them from T

Reply:
Done